soc/xtensa/intel_adsp/common/cavs_ipc.c

```diff
@@ -0,0 +1,164 @@
+		bool done = true;
+
+		if (devdata->handle_message != NULL) {
+			uint32_t msg = regs->tdr & ~CAVS_IPC_BUSY;
+			uint32_t ext = regs->tdd;
+
+			done = devdata->handle_message(dev, devdata->handler_arg,
+						       msg, ext);
+		}
+
+		regs->tdr = CAVS_IPC_BUSY;
+		if (done && !IS_ENABLED(CONFIG_SOC_SERIES_INTEL_CAVS_V15)) {
+			regs->tda = CAVS_IPC_DONE;
+		}
+	}
+
+	/* Same signal, but on different bits in 1.5 */
+	bool done = IS_ENABLED(CONFIG_SOC_SERIES_INTEL_CAVS_V15) ?
+		(regs->idd & CAVS_IPC_IDD15_DONE) : (regs->ida & CAVS_IPC_DONE);
+
+	if (done) {
+		if (devdata->done_notify != NULL) {
+			devdata->done_notify(dev, devdata->done_arg);
+		}
+		k_sem_give(&devdata->sem);
+		if (IS_ENABLED(CONFIG_SOC_SERIES_INTEL_CAVS_V15)) {
+			regs->idd = CAVS_IPC_IDD15_DONE;
+		} else {
+			regs->ida = CAVS_IPC_DONE;
+		}
+	}
+
+	k_spin_unlock(&devdata->lock, key);
+}
+
+int cavs_ipc_init(const struct device *dev)
+{
+	struct cavs_ipc_data *devdata = dev->data;
+	const struct cavs_ipc_config *config = dev->config;
+
+	memset(devdata, 0, sizeof(*devdata));
+
+	/* ACK any latched interrupts (including TDA to clear IDA on
+	 * the other side!), then enable.
+	 */
+	config->regs->tdr = CAVS_IPC_BUSY;
+	if (IS_ENABLED(CONFIG_SOC_SERIES_INTEL_CAVS_V15)) {
+		config->regs->idd = CAVS_IPC_IDD15_DONE;
+	} else {
+		config->regs->ida = CAVS_IPC_DONE;
+		config->regs->tda = CAVS_IPC_DONE;
+	}
+	config->regs->ctl |= (CAVS_IPC_CTL_IDIE | CAVS_IPC_CTL_TBIE);
+	return 0;
+}
+
+void cavs_ipc_complete(const struct device *dev)
+{
+	const struct cavs_ipc_config *config = dev->config;
+
+	config->regs->tda = CAVS_IPC_DONE;
+}
+
+bool cavs_ipc_is_complete(const struct device *dev)
+{
+	const struct cavs_ipc_config *config = dev->config;
+
+	return (config->regs->idr & CAVS_IPC_BUSY) == 0;
+}
+
+bool cavs_ipc_send_message(const struct device *dev,
+			   uint32_t data, uint32_t ext_data)
+{
+	const struct cavs_ipc_config *config = dev->config;
+	struct cavs_ipc_data *devdata = dev->data;
+	k_spinlock_key_t key = k_spin_lock(&devdata->lock);
+
+	if ((config->regs->idr & CAVS_IPC_BUSY) != 0) {
+		k_spin_unlock(&devdata->lock, key);
+		return false;
+	}
+
+	k_sem_init(&devdata->sem, 0, 1);
+	config->regs->idd = ext_data;
+	config->regs->idr = data | CAVS_IPC_BUSY;
+	k_spin_unlock(&devdata->lock, key);
+	return true;
+}
+
+bool cavs_ipc_send_message_sync(const struct device *dev,
+				uint32_t data, uint32_t ext_data,
+				k_timeout_t timeout)
+{
+	struct cavs_ipc_data *devdata = dev->data;
+
+	bool ret = cavs_ipc_send_message(dev, data, ext_data);
+
+	if (ret) {
+		k_sem_take(&devdata->sem, timeout);
+	}
+	return ret;
+}
+
+#if DT_NODE_EXISTS(CAVS_HOST_DTNODE)
+static int dt_init(const struct device *dev)
+{
+	const int irq = DT_IRQN(CAVS_HOST_DTNODE);
+
+	/* FIXME: the irq expression above is clearly a constant, but
+	 * using it with IRQ_CONNECT here produces a wild preprocessor
+	 * error.  Why?
+	 */
+	irq_connect_dynamic(irq, 0, (void *)z_cavs_ipc_isr, dev, 0);
```

Comment:
No, that's how I had it.  It didn't build.  Frankly it doesn't matter for these platforms since dynamic IRQs are architectural, but something to investigate.

Reply:
Is this with the xtensa compiler in the sdk or with the proprietary job? worked fine for me with the zephyr sdk compiler as I wrote it. I just built the board test with it.

Reply:
It was just the SDK.  I'll go back and check, it's entirely possible I messed something up.  I have minimal patience for devicetree burps, to be perfectly honest.

Reply:
Turned out that it didn't like the left typecast in the ISR parameter, because it wants to use it to glue up a symbol name.   Fixed.